nssdb/nssdb.go

```diff
@@ -0,0 +1,457 @@
+package nssdb
+
+import (
+	"context"
+	"crypto/sha1" //nolint:gosec // NSS uses sha1
+	"database/sql"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	// enable sql driver
+	_ "modernc.org/sqlite"
+)
+
+var columnNames = make(map[string]string, len(columns))
+
+func init() {
+	for k, v := range columns {
+		columnNames[v] = k
+	}
+}
+
+type NSSDB struct {
+	Key      *sql.DB
+	Cert     *sql.DB
+	columns  []string
+	colTable map[string]bool
+	// aka "intermediate key", this is derived from the user-provided password and
+	// the "global salt" in the metaData. This is used as input to pbkdf2 for all
+	// encryption and signing operations.
+	passKey       []byte
+	emptyPassword bool
+}
+
+func (db *NSSDB) Close() error {
+	if err := db.Key.Close(); err != nil {
+		return fmt.Errorf("close key db: %w", err)
+	}
+	if err := db.Cert.Close(); err != nil {
+		return fmt.Errorf("close cert db: %w", err)
+	}
+	return nil
+}
+
+// New opens connections to the cert9 and key4 sqlite databases in the provided
+// directory. It defaults to the current directory if not set. The password
+// argument is not required if the NSS database was created with the
+// --empty-password flag.
+func New(dir string, pw []byte) (*NSSDB, error) {
+	if dir == "" {
+		dir = "."
+	}
+	keyfile := filepath.Join(dir, "key4.db")
+	certfile := filepath.Join(dir, "cert9.db")
+
+	if _, err := os.Stat(keyfile); err != nil {
+		return nil, fmt.Errorf("no nss database found in %q", dir)
+	}
+	if _, err := os.Stat(certfile); err != nil {
+		return nil, fmt.Errorf("no nss database found in %q", dir)
+	}
+
+	keydb, err := sql.Open("sqlite", "file:"+keyfile)
+	if err != nil {
+		return nil, fmt.Errorf("open keydb %q: %w", keyfile, err)
+	}
+
+	certdb, err := sql.Open("sqlite", "file:"+certfile)
+	if err != nil {
+		return nil, fmt.Errorf("open certdb %q: %w", certfile, err)
+	}
+
+	// For backward and forward compatibility we need to know what columns are in
+	// the nssPrivate and nssPublic tables. They share the same schema.
+	rows, err := certdb.Query("SELECT * FROM nssPublic LIMIT 0")
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	var dbcolumns []string
+	colTable := map[string]bool{}
+	cols, err := rows.ColumnTypes()
+	if err != nil {
+		return nil, err
+	}
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+	for _, col := range cols {
+		name := col.Name()
+		if _, ok := columnNames[name]; !ok {
+			// ignore any unknown columns
+			continue
+		}
+		dbcolumns = append(dbcolumns, name)
+		colTable[name] = true
+	}
+
+	var globalSalt []byte
+	err = keydb.QueryRow(`SELECT item1 FROM metaData WHERE id = "password"`).Scan(&globalSalt)
```

Comment:
Although it works, SQL usually uses single quotes for strings.

Reply:
```suggestion
	err = keydb.QueryRow(`SELECT item1 FROM metaData WHERE id = 'password'`).Scan(&globalSalt)
```

Reply:
2168975